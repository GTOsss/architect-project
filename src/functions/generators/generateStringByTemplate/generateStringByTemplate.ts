import { createInterpolationRegExp, parseAllInterpolationMarks } from '../../parsers';
import { ParserContextEnum } from '../../parsers/parseTemplateFiles';
import { generateContentByParsedTemplate } from '../generateContentByParsedTemplate/generateContentByParsedTemplate';

const intrConfig = { itrStart: '{{', itrEnd: '}}' };

/**
 * Utility function for generate text content by templates.
 *
 * @example
 * const simpleTemplate = `const {{methodName}} = ({{arg1}}, {{arg2}}) => {{arg1}} + {{arg2}}`;
 * const templateParams = { methodName: 'testMethod', arg1: 'firstArg', arg2: 'secondArg' };
 *
 * const result = generateStringByTemplate(simpleTemplate, templateParams);
 * // const testMethod = (firstArg, secondArg) => firstArg + secondArg
 * */
export const generateStringByTemplate = (template: string, params: Record<string, any>) => {
  const intrRegExp = createInterpolationRegExp(intrConfig);

  const parsed = parseAllInterpolationMarks(template, intrRegExp, intrConfig, ParserContextEnum.fileContent);
  return generateContentByParsedTemplate({ parsed, content: template }, params, true);
};
