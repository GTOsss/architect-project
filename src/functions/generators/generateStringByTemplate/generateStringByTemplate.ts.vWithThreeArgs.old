import { createInterpolationRegExp, parseAllInterpolationMarks } from '../../parsers';
import { ParserContextEnum } from '../../parsers/parseTemplateFiles';
import { generateContentByParsedTemplate } from '../generateContentByParsedTemplate/generateContentByParsedTemplate';

const defaultInterpolationConfig = { itrStart: '{{', itrEnd: '}}' };

type IntrFileConfig<IS extends string, IE extends string> = {
  itrStart: IS;
  itrEnd: IE;
};

type ExtractTemplateKeys<
  S extends string,
  IS extends string,
  IE extends string,
> = S extends `${infer _Start}${IS}${infer Key}${IE}${infer Rest}` ? Key | ExtractTemplateKeys<Rest, IS, IE> : never;

type GenerateStringByTemplateParams<T extends string, IS extends string, IE extends string> = Record<
  ExtractTemplateKeys<T, IS, IE>,
  string
>;

/**
 * Utility function for generating text content by templates.
 *
 * @example
 * const simpleTemplate = `const {{methodName}} = ({{arg1}}, {{arg2}}) => {{arg1}} + {{arg2}}`;
 * const templateParams = { methodName: 'testMethod', arg1: 'firstArg', arg2: 'secondArg' };
 *
 * const result = generateStringByTemplate(simpleTemplate, templateParams);
 * // const testMethod = (firstArg, secondArg) => firstArg + secondArg
 * */
export function generateStringByTemplate<T extends string, IS extends string = '{{', IE extends string = '}}'>(
  template: T,
  params: GenerateStringByTemplateParams<T, IS, IE>,
  interpolationConfig?: IntrFileConfig<IS, IE>,
) {
  const intrConfig = interpolationConfig || defaultInterpolationConfig;
  const intrRegExp = createInterpolationRegExp(intrConfig);

  const parsed = parseAllInterpolationMarks(template, intrRegExp, intrConfig, ParserContextEnum.fileContent);
  return generateContentByParsedTemplate({ parsed, content: template }, params as unknown as Record<string, any>, true);
}
