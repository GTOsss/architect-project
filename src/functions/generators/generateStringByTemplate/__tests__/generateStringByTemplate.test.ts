import { generateStringByTemplate } from '../generateStringByTemplate';

describe('generateStringByTemplate', () => {
  it('simple', () => {
    const simpleTemplate = `const {{methodName}} = ({{arg1}}, {{arg2}}) => {{arg1}} + {{arg2}}`;
    const templateParams = { methodName: 'testMethod', arg1: 'firstArg', arg2: 'secondArg' };

    const result = generateStringByTemplate(simpleTemplate, templateParams);
    expect(result).toMatchSnapshot();
  });
});
