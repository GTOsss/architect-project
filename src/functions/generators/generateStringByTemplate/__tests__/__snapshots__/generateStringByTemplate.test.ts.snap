// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateStringByTemplate simple 1`] = `"const testMethod = (firstArg, secondArg) => firstArg + secondArg"`;
